bip-0141.mediawiki

```diff
@@ -58,22 +58,15 @@ A non-witness program (defined hereinafter) txin MUST be associated with an empt
 
 === Commitment structure ===
 
-A new block rule is added which requires a commitment to the <code>wtxid</code>. The <code>wtxid</code> of coinbase transaction is assumed to be <code>0x0000....0000</code>.
+The <code>scriptSig</code> of coinbase transaction gets a second number push ("offset"), immediately following the block height push of [[bip-0034.mediawiki|BIP34]]. Same as the block height push, format of the number is "serialized CScript". The first byte is number of bytes in the number, and the following bytes are little-endian representation of the number.
 
-A witness root hash is calculated with all those <code>wtxid</code> as leaves, in a way similar to the hashMerkleRoot in the block header.
+The "offset" is a byte offset inside the coinbase transaction serialized in the traditional format (without <code>marker</code>, <code>flag</code>, or <code>witness</code> described in the previous section). It points to a 32-byte value ("witness commitment").
 
-The commitment is recorded in a scriptPubKey of the coinbase transaction. It must be at least 38 bytes, with the first 6-byte of <code>0x6a24aa21a9ed</code>, that is:
-  
-   1-byte - OP_RETURN (0x6a)
-   1-byte - Push the following 36 bytes (0x24)
-   4-byte - Commitment header (0xaa21a9ed)
-  32-byte - Commitment hash: Double-SHA256(witness root hash|witness nonce)
-  
-  39th byte onwards: Optional data with no consensus meaning
-  
-and the coinbase's input's witness must consist of a single 32-byte array for the witness nonce.
+For each stack item in the coinbase witness, a double SHA256 is calculated. The "witness commitment" is the root of a Merkle tree, whose leaves are all these hashes.
 
-If there are more than one scriptPubKey matching the pattern, the one with highest output index is assumed to be the commitment.
+Item 0 of the coinbase witness stack is the "wtxid commitment". The "wtxid commitment" must be exactly 32 bytes. It is the root of a Merkle tree, whose leaves are all those <code>wtxid</code> in the block, calculated in a way similar to the hashMerkleRoot in the block header. The <code>wtxid</code> of coinbase transaction is assumed to be <code>0x0000....0000</code>.
```

Comment:
Item 0 mean top of the stack or bottom of the stack ? I assume bottom, but maybe it can be clarified.